extensions/amp-skimlinks/0.1/tracking.js

```diff
@@ -0,0 +1,216 @@
+import {CustomEventReporterBuilder} from '../../../src/extension-analytics.js';
+import {generatePageImpressionId} from './utils';
+
+import {PLATFORM_NAME, XCUST_ATTRIBUTE_NAME} from './constants';
+
+
+
+import {
+  LINKS_IMPRESSIONS_TRACKING_URL,
+  NA_CLICK_TRACKING_URL,
+  PAGE_IMPRESSION_TRACKING_URL,
+} from './constants';
+
+
+export default class Tracking {
+  /**
+   * Use tracking instance to track page impressions,
+   * link impressions and non-affiliated clicks.
+   * @param {*} element
+   * @param {*} skimOptions
+   */
+  constructor(element, skimOptions) {
+    this.tracking_ = skimOptions.tracking;
+
+    this.trackingInfo_ = {
+      pubcode: skimOptions.pubcode,
+      // https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md
+      pageUrl: 'CANONICAL_URL',
+      referrer: 'DOCUMENT_REFERRER',
+      timezone: 'TIMEZONE',
+      pageImpressionId: generatePageImpressionId(),
+      customTrackingId: skimOptions.customTrackingId,
+      guid: null,
+    };
+
+    this.analytics_ = this.setupAnalytics_(element);
+  }
+
+  /**
+   *
+   * @param {*} element
+   */
+  setupAnalytics_(element) {
+    // Analytics are not ready until CommonSignals.LOAD_START is triggered.
+    const analyticsBuilder = new CustomEventReporterBuilder(element);
+    analyticsBuilder.track('page-impressions', PAGE_IMPRESSION_TRACKING_URL);
+    analyticsBuilder.track('link-impressions', LINKS_IMPRESSIONS_TRACKING_URL);
+    analyticsBuilder.track('non-affiliate-click', NA_CLICK_TRACKING_URL);
+
+    const analytics = analyticsBuilder.build();
+    // Overwrite config manually since CustomEventReporterBuilder doesn't
+    // support optional config.
+    // TODO: add optional config param to .build() so we don't need to mutate
+    // a private property from outside.
+    analytics.config_.transport = {beacon: true};
+
+    return analytics;
+  }
+
+  /**
+   *
+   */
+  getTrackingInfo() {
+    return this.trackingInfo_;
+  }
+
+  /**
+   * Update tracking info
+   * @param {*} newInfo
+   */
+  setTrackingInfo(newInfo) {
+    Object.assign(this.trackingInfo_, newInfo);
+  }
+
+  /**
+   * Send Page impression and link impressions
+   * @param {*} anchorStatusMap
+   * @param {*} startTime
+   */
+  sendImpressionTracking(anchorStatusMap, startTime) {
+    if (!this.tracking_) {
```

Comment:
I guess this happens when tracking is disabled, perhaps a more readable way to have this is having a `shouldNotTrack_` (or `willNotTrack_`) method that reads the `this.tracking_` (Or any other future condition).

Not too fussed about it, just think it makes it easier to read/understand